src/crush/CrushWrapper.cc

```diff
@@ -952,13 +963,31 @@ int CrushWrapper::verify_upmap(CephContext *cct,
                           << dendl;
           }
         }
+        ldout(cct, 10) << __func__ << " osds_by_parent " << osds_by_parent << dendl;
+        ceph_assert(root_bucket < 0);
         for (auto i : osds_by_parent) {
           if (i.second.size() > 1) {
             lderr(cct) << __func__ << " multiple osds " << i.second
                        << " come from same failure domain " << i.first
                        << dendl;
             return -EINVAL;
           }
+          if (subtree_contains(root_bucket, i.first)) {
+            real_rep++;
+          }
+        }
+        if (real_rep != numrep) {
+          lderr(cct) << __func__ << " expected " << numrep << " items in bucket " << root_bucket
+                     << " real " << real_rep << dendl;
+          return -EINVAL;
+        }
+        // validate the osd's in subtree
+        for (int c = 0; cursor < up.size() && c < numrep; ++cursor, ++c) {
```

Comment:
hmm, this looks the wrong way to fix the problem to me.

actually I think you should instead verify that:
- we have exactly  2 up osds from datacenter-1
- we have exactly 2 up osds from datacenter-2
- there is no more remaining up osds...

Reply:
- we have exactly 2 up (up[0] and up[1]) osds from datacenter-1

- we have exactly 2 up (up[2] and up[3]) osds from datacenter-2

- there is no more remaining up osds...

Reply:
ahh, i got you now